### PR TITLE
Define a mechanism for collecting per-layer metrics during a forward pass

### PR DESCRIPTION
**Pull request by nschank**
# What does this PR do ?

* Defines an interfaces for collecting metrics from individual layers on a forward pass, as well as for reporting those metrics somewhere else (such as WandB, MLFlow, Prometheus...).
* Provides an implementation that allows users to define configurable 'rules' which capture metrics on a per-`Module`-type basis
* Plugs in the `collector` object in a few key modules to demonstrate usage (can expand usages further on request)

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [X] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by nschank** [comment]
Note that this change is being coordinated with @yashaswikarnati as the ReflectionAI PoC

**Comment by yashaswikarnati** [comment]
Thanks for the great contributions. This looks good to me, just have 2 minor comments/requests

1) Could we add a unit test show casing how a rule based aggregator can be used to report metrics for a particular moe layers such as MoE layer (just as an example).
2) could we add a brief doc in docs/source/api-guide on how this can be used for different layers and may be point to the above unit test

File: megatron/core/metrics/every_n_reporter.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from typing import final, override
+
+from . import metrics, reporter
```
Comment (Phlip79) [inline megatron/core/metrics/every_n_reporter.py:3]:
```suggestion
from megatron.core.metrics import metrics, reporter
```

File: megatron/core/metrics/forwarding_collector.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+from typing import TYPE_CHECKING, override
+
+from . import collector
```
Comment (Phlip79) [inline megatron/core/metrics/forwarding_collector.py:3]:
```suggestion
from megatron.core.metrics import collector
```

File: megatron/core/metrics/reporter.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import abc
+from typing import final, override
+
+from . import metrics
```
Comment (Phlip79) [inline megatron/core/metrics/reporter.py:4]:
```suggestion
from megatron.core.metrics import metrics
```

File: megatron/core/metrics/rule_based_aggregator.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+import abc
+import collections
+from typing import Mapping, final, override
+
+import torch
+
+from megatron.core import process_groups_config
+
+from . import collector, metrics, reporter
```
Comment (Phlip79) [inline megatron/core/metrics/rule_based_aggregator.py:16]:
```suggestion
from megatron.core.metrics import collector, metrics, reporter
```

File: megatron/core/transformer/attention.py [hunk 5]
```diff
@@ -148,6 +149,7 @@ def __init__(
         attention_type: str,
         cp_comm_type: str = None,
         pg_collection: ProcessGroupCollection = None,
+        metric_collector: collector.MetricCollector = (collector.NoopMetricCollector()),
```
Comment (Phlip79) [inline megatron/core/transformer/attention.py:152]:
```suggestion
        metric_collector: collector.MetricCollector = collector.NoopMetricCollector(),
```
